sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +408,284 @@ class GpuSequenceMeta(
-              pred.ifElse(one, negativeOne)
+  def computeArraySizes(
+      start: ColumnVector,
+      stop: ColumnVector,
+      step: ColumnVector): ColumnVector = {
+    // Keep the same requirement with Spark:
+    //  (step > 0 && start <= stop) || (step < 0 && start >= stop) || (step == 0 && start == stop)
+    withResource(Scalar.fromByte(0.toByte)) { zero =>
+      withResource(new Table(start, stop)) { startStopTable =>
+        // (step > 0 && start <= stop)
+        withResource(step.greaterThan(zero)) { positiveStep =>
+          withResource(startStopTable.filter(positiveStep)) { upTbl =>
+            withResource(upTbl.getColumn(0).lessOrEqualTo(upTbl.getColumn(1))) { allUp =>
+              require(isAllTrue(allUp), "Illegal sequence boundaries: step > 0 but start > stop")
             }
           }
         }
-        val size = closeOnExcept(step) { _ =>
-          withResource(difference.abs()) { absDifference =>
-            absDifference.add(one)
+        // (step < 0 && start >= stop)
+        withResource(step.lessThan(zero)) { negativeStep =>
+          withResource(startStopTable.filter(negativeStep)) { downTbl =>
+            withResource(downTbl.getColumn(0).greaterOrEqualTo(downTbl.getColumn(1))) { allDown =>
+              require(isAllTrue(allDown),
+                "Illegal sequence boundaries: step < 0 but start < stop")
+            }
+          }
+        }
+        // (step == 0 && start == stop)
+        withResource(step.equalTo(zero)) { zeroStep =>
+          withResource(startStopTable.filter(zeroStep)) { equalTbl =>
+            withResource(equalTbl.getColumn(0).equalTo(equalTbl.getColumn(1))) { allEqual =>
+              require(isAllTrue(allEqual),
+                "Illegal sequence boundaries:  step == 0 but start != stop")
+            }
           }
         }
-        Seq(size, step)
       }
     }
-  }
-
-  override def doColumnar(start: GpuColumnVector, stop: GpuColumnVector): ColumnVector = {
-    withResource(calculateSizeAndStep(start.getBase, stop.getBase, start.dataType())) { ret =>
-      ColumnVector.sequence(start.getBase, ret(0), ret(1))
+    // Spark's algorithm to get the length (aka size)
+    // ``` Scala
+    //   size = if (start == stop) 1L else 1L + (stop.toLong - start.toLong) / estimatedStep.toLong
+    //   require(size <= MAX_ROUNDED_ARRAY_LENGTH,
+    //       s"Too long sequence: $len. Should be <= $MAX_ROUNDED_ARRAY_LENGTH")
+    //   size.toInt
+    // ```
+    val sizeAsLong = withResource(Scalar.fromLong(1L)) { one =>
+      val diff = withResource(stop.castTo(DType.INT64)) { stopAsLong =>
+        withResource(start.castTo(DType.INT64)) { startAsLong =>
+          stopAsLong.sub(startAsLong)
+        }
+      }
+      // actualSize = 1L + (stop.toLong - start.toLong) / estimatedStep.toLong
+      val actualSize = withResource(diff) { _ =>
+        withResource(step.castTo(DType.INT64)) { stepAsLong =>
```

Comment:
`stepAsLong` should be closed immediately after computing `quotient`.

Reply:
done